cmd/kubeadm/app/master/tokens.go

```diff
@@ -28,34 +28,32 @@ import (
 	"k8s.io/kubernetes/pkg/util/uuid"
 )
 
-func generateTokenIfNeeded(s *kubeadmapi.Secrets) error {
-	ok, err := kubeadmutil.UseGivenTokenIfValid(s)
-	// TODO(phase1+) @krousey: I know it won't happen with the way it is currently implemented, but this doesn't handle case where ok is true and err is non-nil.
-	if !ok {
-		if err != nil {
-			return err
-		}
-		err = kubeadmutil.GenerateToken(s)
-		if err != nil {
-			return err
-		}
-		fmt.Printf("<master/tokens> generated token: %q\n", s.GivenToken)
-	} else {
+func generateTokenIfNeeded(d *kubeadmapi.TokenDiscovery) error {
+	ok, err := kubeadmutil.IsTokenValid(d)
+	if err != nil {
+		return err
+	}
+	if ok {
 		fmt.Println("<master/tokens> accepted provided token")
+		return nil
+	}
+	if err := kubeadmutil.GenerateToken(d); err != nil {
+		fmt.Printf("<master/tokens> generated token: %q\n", kubeadmutil.BearerToken(d))
+		return nil
+	} else {
+		return err
```

Comment:
I will have some merge conflicts to deal with here on the token cli PR but we should probably merge this one first, and then I will update on my end.